### PR TITLE
[Draft] Off-hand

### LINKED ISSUE #158: Implement off-hand slot in hot bar (press F to exchange items)
follow up of #84

### PR DESCRIPTION
**Pull request by Pandapip1**
Fixes #158

- [x] Press F to ~~pay respects~~ swap main hand and off-hand
- [x] Display off-hand in Hotbar
- [x] Display offhand slot
- [x] Hide off-hand Hotbar slot when there's nothing in it
- [ ] Testing

### CODE REVIEW

**Comment by Pandapip1** [comment]
Status Update: Math is hard and I'm still working out the positions.

**Comment by Pandapip1** [comment]
Testing.

**Comment by Pandapip1** [comment]
Okay, *now* I'm testing.

File: lib/hotbar.js [hunk 1]
```diff
@@ -78,8 +100,8 @@ class HotBar extends LitElement {
       left: 50%;
       height: calc(256px * 4);
       width: calc(256px * 4);
-      transform: translate(calc(-182px * 2), calc(-22px * 4));
-      clip-path: inset(0px calc(74px * 4) calc(234px * 4) 0px);
+      transform: translate(-35.555%, -8.594%);
+      clip-path: inset(0% 28.906% 91.406% 0%);
```
Comment (Saiv46) [inline lib/hotbar.js:104]: Please avoid making _magic numbers_, using `calc()` will improve readability a much.

File: lib/hotbar.js [hunk 2]
```diff
@@ -20,26 +20,48 @@ class HotBar extends LitElement {
...
       slotImage.style['background-position-x'] = `-${sprite.x * 2}px`
       slotImage.style['background-position-y'] = `-${sprite.y * 2}px`
       slotImage.innerHTML = item?.count > 1 ? item.count : ''
-    }
+      if (key === 'offhand') {
+        const offhandImg = this.shadowRoot.getElementById('hotbar-offhand-image')
+        if (item) offhandImg.show()
+        else offhandImg.hide()
+      }
+    })
```
Comment (Saiv46) [inline lib/hotbar.js:64]: Why use `Object.keys`?
```suggestion
    function updateSlot (key, item) {
      const sprite = item ? invsprite[item.name] : invsprite.air
      const slotImage = this.shadowRoot.getElementById('hotbar-' + key)
      slotImage.style['background-position-x'] = `-${sprite.x * 2}px`
      slotImage.style['background-position-y'] = `-${sprite.y * 2}px`
      slotImage.innerHTML = item?.count > 1 ? item.count : ''
      if (key === 'offhand') {
        const offhandImg = this.shadowRoot.getElementById('hotbar-offhand-image')
        if (item) offhandImg.show()
        else offhandImg.hide()
      }
    }
    for (let i = 0; i < 9; i++) {
      updateSlot(i, this.bot.inventory.slots[this.bot.inventory.hotbarStart + i])
    }
    updateSlot('offhand', this.bot.getEquipmentDestSlot('off-hand'))
```

Comment (Pandapip1) [reply lib/hotbar.js:64]: Yea, this is a better way of doing it.

File: lib/hotbar.js (hunk 1 above)
Comment (Pandapip1) [reply lib/hotbar.js:104]: The problem here is that the inset numbers really truly are magic numbers that are specific to the widgets image. I'll see what I can do about adding calc to the transforms.

Comment (Pandapip1) [reply lib/hotbar.js]: I've made the transforms using `calc()`. The insets are by nature magic numbers so I've made comments explaining how to make them.

**Comment by Pandapip1** [comment]
Testing url: https://chatfade-pjs-test.pandapip1.com/

(Note: I just reused my URL for #236)

**Review by host008 — APPROVED** [review]

**Comment by rom1504** [comment]
Could you rebase on master ?

**Comment by Pandapip1** [comment]
This is somewhat (read: very) broken. I'm going to close this but leave the code there in case anyone wants to try to redo it.

**Pull request closed without merging**